mixing/mixpool/mixpool.go

```diff
@@ -0,0 +1,1523 @@
+			bytes.Equal(e.msg.Sid(), msg.Sid()) {
+			return nil, fmt.Errorf("message %v by identity %x "+
+				"reuses run number %d in session %x, "+
+				"conflicting with already accepted message %v",
+				hash, *id, run, msg.Sid(), prevHash)
+		}
+		if e.msgtype == msgtypeKE && e.msg.GetRun() == run &&
+			bytes.Equal(e.msg.Sid(), msg.Sid()) {
+			haveKE = true
+		}
+	}
+	// Save as an orphan if their KE is not (yet) known.
+	if !haveKE {
+		orphansByID := p.orphansByID[*id]
+		if _, ok := orphansByID[hash]; ok {
+			// Already an orphan.
+			return nil, nil
+		}
+		if orphansByID == nil {
+			orphansByID = make(map[chainhash.Hash]mixing.Message)
+			p.orphansByID[*id] = orphansByID
+		}
+		p.orphans[hash] = &orphan{
+			message:  msg,
+			accepted: time.Now(),
+		}
+		orphansByID[hash] = msg
+
+		// TODO: Consider return an error containing the unknown
+		// messages, so they can be getdata'd.
+		return nil, nil
+	}
+
+	ses := p.sessions[sid]
+	if ses == nil {
+		return nil, fmt.Errorf("%s %s belongs to unknown session %x",
+			msgtype, &hash, &sid)
+	}
+
+	err = p.acceptEntry(msg, msgtype, &hash, id, ses)
+	if err != nil {
+		return nil, err
+	}
+	return []mixing.Message{msg}, nil
+}
+
+// removePR removes a pair request message and all other messages and sessions
+// that the peer sent and was involved in.
+func (p *Pool) removePR(pr *wire.MsgMixPairReq, reason string) {
+	prHash := pr.Hash()
+
+	log.Debugf("Removing %s PR %s by %x", reason, prHash, pr.Identity[:])
+
+	delete(p.prs, prHash)
+	for _, hash := range p.messagesByIdentity[pr.Identity] {
+		e, ok := p.pool[hash]
+		if !ok {
+			continue
+		}
+		ke, ok := e.msg.(*wire.MsgMixKeyExchange)
+		if ok {
+			p.removeSession(ke.SessionID, nil, false)
+		}
+		delete(p.pool, hash)
+	}
+	delete(p.messagesByIdentity, pr.Identity)
+	delete(p.latestKE, pr.Identity)
+	for orphanHash := range p.orphansByID[pr.Identity] {
+		delete(p.orphans, orphanHash)
+	}
+	delete(p.orphansByID, pr.Identity)
+	for i := range pr.UTXOs {
+		delete(p.outPoints, pr.UTXOs[i].OutPoint)
+	}
+}
+
+func (p *Pool) acceptPR(pr *wire.MsgMixPairReq, hash *chainhash.Hash, id *idPubKey) (accepted *wire.MsgMixPairReq, err error) {
+	switch {
+	case len(pr.UTXOs) == 0: // Require at least one utxo.
+		return nil, ruleError(ErrMissingUTXOs)
+	case pr.MessageCount == 0: // Require at least one mixed message.
+		return nil, ruleError(ErrInvalidMessageCount)
+	case pr.InputValue < int64(pr.MessageCount)*pr.MixAmount:
+		return nil, ruleError(ErrInvalidTotalMixAmount)
+	case pr.Change != nil:
+		if isDustAmount(pr.Change.Value, p2pkhv0PkScriptSize, feeRate) {
+			return nil, ruleError(ErrChangeDust)
+		}
+		if !stdscript.IsPubKeyHashScriptV0(pr.Change.PkScript) &&
+			!stdscript.IsScriptHashScriptV0(pr.Change.PkScript) {
+			return nil, ruleError(ErrInvalidScript)
+		}
+	}
+
+	// Check that expiry has not been reached, nor that it is too far
+	// into the future.  This limits replay attacks.
+	_, curHeight := p.blockchain.BestHeader()
```

Comment:
I added a comment on the definition of `BestHeader` in the interface, but, if the plan was to use `internal/blockchain/BlockChain.BestHeader` to satisfy the interface, that would result in unexpected behavior here I believe, because it looks to me like it is expecting it to be the height of the current best chain tip that has been validated and that is **NOT** what `internal/blockchain/BlockChain.BestHeader` does.

For the height of the current best chain tip that has been validated , you would need to get it from `BlockChain.BestSnapshot().Height`